apps/meteor/app/livechat/server/lib/ContactMerger.ts

```diff
@@ -0,0 +1,281 @@
+				return false;
+			}
+
+			return true;
+		});
+
+		const newPhones = ContactMerger.getFieldValuesByType(newFields, 'phone');
+		const newEmails = ContactMerger.getFieldValuesByType(newFields, 'email');
+		const newChannels = ContactMerger.getFieldValuesByType(newFields, 'channel');
+		const newNamesOnly = ContactMerger.getFieldValuesByType(newFields, 'name');
+		const newCustomFields = newFields.filter(({ type }) => type.startsWith('customFields.')) as CustomFieldAndValue[];
+		// Usernames are ignored unless the contact has no other name
+		const newUsernames = !contact.name && !newNamesOnly.length ? ContactMerger.getFieldValuesByType(newFields, 'username') : [];
+
+		const dataToSet: DeepWritable<UpdateFilter<ILivechatContact>['$set']> = {};
+
+		// Names, Managers and Custom Fields need are set as conflicting fields if the contact already has them
+		const newNames = [...newNamesOnly, ...newUsernames];
+		const newManagers = ContactMerger.getFieldValuesByType(newFields, 'manager')
+			.map((manager) => {
+				if ('id' in manager) {
+					return manager.id;
+				}
+				return merger.getManagerId(manager);
+			})
+			.filter((id) => Boolean(id));
+
+		if (newNames.length && !contact.name) {
+			const firstName = newNames.shift();
+			if (firstName) {
+				dataToSet.name = firstName;
+			}
+		}
+
+		if (newManagers.length && !contact.contactManager) {
+			const firstManager = newManagers.shift();
+			if (firstManager) {
+				dataToSet.contactManager = firstManager;
+			}
+		}
+
+		const customFieldsPerName = new Map<string, CustomFieldAndValue[]>();
+		for (const customField of newCustomFields) {
+			if (!customFieldsPerName.has(customField.type)) {
+				customFieldsPerName.set(customField.type, []);
+			}
+			customFieldsPerName.get(customField.type)?.push(customField);
+		}
+
+		const customFieldConflicts: CustomFieldAndValue[] = [];
+
+		for (const [key, customFields] of customFieldsPerName) {
+			const fieldName = key.replace('customFields.', '');
+
+			// If the contact does not have this custom field yet, save the first value directly to the contact instead of as a conflict
+			if (!contact.customFields?.[fieldName]) {
+				const first = customFields.shift();
+				if (first) {
+					dataToSet[key] = first.value;
+				}
+			}
+
+			customFieldConflicts.push(...customFields);
+		}
+
+		const allConflicts: ILivechatContactConflictingField[] = [
+			...newNames.map((name): ILivechatContactConflictingField => ({ field: 'name', value: name })),
+			...newManagers.map((manager): ILivechatContactConflictingField => ({ field: 'manager', value: manager as string })),
+			...customFieldConflicts.map(({ type, value }): ILivechatContactConflictingField => ({ field: type, value })),
+		];
+
+		if (allConflicts.length) {
+			dataToSet.hasConflicts = true;
+		}
+
+		// Phones, Emails and Channels are simply added to the contact's existing list
+		const dataToAdd: UpdateFilter<ILivechatContact>['$addToSet'] = {
```

Comment:
maybe that orm thing would have been of help here to avoid the verbose code. Maybe for a future version, not sure if we should create a task for it?

Reply:
I'll add this to our CORE board later so as to not block the project